File: Data Structures & Algorithms/two-integer-sum-ii/submission-2.py
class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        l, r = 0, (len(numbers) - 1)
        while l < r:
            if target == (numbers[l] + numbers[r]):
                return [l + 1, r + 1]
            if target > (numbers[l] + numbers[r]):
                l += 1
            if target < (numbers[l] + numbers[r]):
                r -= 1
            

        

File: Data Structures & Algorithms/target-sum/submission-2.java
class Solution {
    private Map<String, Integer> memo = new HashMap<>();
    public int findTargetSumWays(int[] nums, int target) {
        return backTrack(nums, target, 0, 0);
    }
    private int backTrack(int[] nums, int target, int index, int currSum){
        String key = index + "," + currSum;
        if (memo.containsKey(key)){
            return memo.get(key);
        }
        if (index == nums.length){
            return currSum == target ? 1 : 0;
        }
        int add = backTrack(nums, target, index + 1, currSum + nums[index]);
        int subtrack = backTrack(nums, target, index + 1, currSum - nums[index]);
        memo.put(key, add + subtrack);
        return memo.get(key);
    }
}
